### PR TITLE
ESQL: Fix tests for HashLookup

### LINKED ISSUE #108045: [CI] HashLookupOperatorTests testSimpleToString failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/lc73f6gbv7i2e/tests/:x-pack:plugin:esql:compute:test/org.elasticsearch.compute.operator.HashLookupOperatorTests/testSimpleToString

**Reproduction line:**
```
./gradlew ':x-pack:plugin:esql:compute:test' --tests "org.elasticsearch.compute.operator.HashLookupOperatorTests.testSimpleToString" -Dtests.seed=6DAC080D5652932 -Dtests.locale=ar-DZ -Dtests.timezone=Australia/Eucla -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
Didn't try

**Failure history:**
[Failure dashboard for `org.elasticsearch.compute.operator.HashLookupOperatorTests#testSimpleToString`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testSimpleToString'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.compute.operator.HashLookupOperatorTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure excerpt:**
```
java.lang.AssertionError: 
Expected: "HashLookup[keys=[foo], hash=PackedValuesBlockHash{groups=[0:LONG], entries=4, size=544b}, mapping=[0]]"
     but: was "HashLookup[keys=[foo], hash=PackedValuesBlockHash{groups=[0:LONG], entries=4, size=552b}, mapping=[0]]"

  at __randomizedtesting.SeedInfo.seed([6DAC080D5652932:4953BD9C6C141B90]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
  at org.elasticsearch.test.ESTestCase.assertThat(ESTestCase.java:2150)
  at org.elasticsearch.compute.operator.AnyOperatorTestCase.testSimpleToString(AnyOperatorTestCase.java:73)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

```

### PR DESCRIPTION
**Pull request by nik9000**
Our hash lookup tests assumed that the size of a hash map was consistent but it's random.... We'd wanted to test randomized toStrings anyway. So this adds support for testing `toString` and `desription`s with hamcrest `Matcher`.

Closes #108045

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/lucene/LuceneSourceOperatorTests.java [hunk 1]
```diff
@@ -93,16 +95,14 @@ private LuceneSourceOperator.Factory simple(DataPartitioning dataPartitioning, i
     }
 
     @Override
-    protected String expectedToStringOfSimple() {
-        assumeFalse("can't support variable maxPageSize", true); // TODO allow testing this
-        return "LuceneSourceOperator[shardId=0, maxPageSize=**random**]";
+    protected Matcher<String> expectedToStringOfSimple() {
+        return matchesRegex("LuceneSourceOperator\\[shardId=0, maxPageSize=\\d+]");
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/lucene/LuceneSourceOperatorTests.java:99]: Why don't we have to escape the closing brackets? Other usages of `matchesRegex` in our code base escape the closing brackets as well.
```suggestion
        return matchesRegex("LuceneSourceOperator\\[shardId=0, maxPageSize=\\d+\\]");
```
Applies to all of the regex added by this PR.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/HashLookupOperatorTests.java [hunk 2]
```diff
@@ -80,12 +82,14 @@ protected Operator.OperatorFactory simple() {
     }
 
     @Override
-    protected String expectedDescriptionOfSimple() {
-        return "HashLookup[keys=[{name=foo, type=LONG, positions=4, size=104b}], mapping=[0]]";
+    protected Matcher<String> expectedDescriptionOfSimple() {
+        return equalTo("HashLookup[keys=[{name=foo, type=LONG, positions=4, size=104b}], mapping=[0]]");
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/HashLookupOperatorTests.java:86]: Why can we assume a fixed size here?

**Review by alex-spies — APPROVED** [review]
LGTM, thanks @nik9000 !

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/lucene/LuceneSourceOperatorTests.java (hunk 1 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/lucene/LuceneSourceOperatorTests.java]: IntelliJ warned me about closing these brackets as "unneeded" and I thought "oh, that makes it slightly more readable....."

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/HashLookupOperatorTests.java (hunk 2 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/HashLookupOperatorTests.java:86]: Woops. That's the whole point....

**Comment by gwbrown** [comment]
If this fixes https://github.com/elastic/elasticsearch/issues/108045, please also revert https://github.com/elastic/elasticsearch/pull/108047!

**Comment by nik9000** [comment]
> If this fixes #108045, please also revert #108047!

Thanks for the ping!

**Pull request merged**